core/src/middleware.rs

```diff
@@ -26,67 +26,149 @@
 
 //! Middleware for `jsonrpsee` servers.
 
-/// Defines a middleware with callbacks during the RPC request life-cycle. The primary use case for
-/// this is to collect timings for a larger metrics collection solution but the only constraints on
-/// the associated type is that it be [`Send`] and [`Copy`], giving users some freedom to do what
-/// they need to do.
+use std::net::SocketAddr;
+
+pub use http::HeaderMap as Headers;
+pub use jsonrpsee_types::Params;
+
+/// Defines a middleware specifically for HTTP requests with callbacks during the RPC request life-cycle.
+/// The primary use case for this is to collect timings for a larger metrics collection solution.
+///
+/// See [`HttpServerBuilder::set_middleware`](../../jsonrpsee_http_server/struct.HttpServerBuilder.html#method.set_middleware) method
+/// for examples.
+pub trait HttpMiddleware: Send + Sync + Clone + 'static {
+	/// Intended to carry timestamp of a request, for example `std::time::Instant`. How the middleware
+	/// measures time, if at all, is entirely up to the implementation.
+	type Instant: std::fmt::Debug + Send + Sync + Copy;
+
+	/// Called when a new JSON-RPC request comes to the server.
+	fn on_request(&self, remote_addr: SocketAddr, headers: &Headers) -> Self::Instant;
+
+	/// Called on each JSON-RPC method call, batch requests will trigger `on_call` multiple times.
+	fn on_call(&self, method_name: &str, params: Params);
```

Comment:
do we want `method kind` here as well i.e, if the call is a "subscription", "method call" or "notification"?

Reply:
Is it easily accessbile? I guess it wouldn't do any harm to have!

Reply:
I don't do it because it requires to lookup the method name in the HashMap.

should be quite easy to do.

Reply:
Mmm, if it's a performance hit to do so (even when no middleware) then I'm happy not at the mo!